app/models/invitation.rb

```diff
@@ -0,0 +1,89 @@
+class Invitation < ApplicationRecord
+  after_initialize :set_defaults
+
+  ACCEPT_STATUS = 'A'.freeze
+  REJECT_STATUS = 'R'.freeze
+  WAITING_STATUS = 'W'.freeze
+
+  belongs_to :to_user, class_name: 'User', foreign_key: 'to_id', inverse_of: false
+  belongs_to :from_user, class_name: 'User', foreign_key: 'from_id', inverse_of: false
+  belongs_to :assignment, class_name: 'Assignment', foreign_key: 'assignment_id'
```

Comment:
Since we have good amount of custom validations in this model, can you use this Custom Validator and test your code?

```
# app/validators/invitation_validator.rb
class InvitationValidator < ActiveModel::Validator
  ACCEPT_STATUS = 'A'.freeze
  REJECT_STATUS = 'R'.freeze
  WAITING_STATUS = 'W'.freeze

  DUPLICATE_INVITATION_ERROR_MSG = "You cannot have duplicate invitations"
  TO_FROM_SAME_ERROR_MSG = "to and from users should be different"
  REPLY_STATUS_ERROR_MSG = "must be present and have a maximum length of 1"
  REPLY_STATUS_INCLUSION_ERROR_MSG = "must be one of #{[ACCEPT_STATUS, REJECT_STATUS, WAITING_STATUS].to_sentence}"

  def validate(record)
    validate_reply_status(record)
    validate_reply_status_inclusion(record)
    validate_duplicate_invitation(record)
    validate_to_from_different(record)
  end

  private

  def validate_reply_status(record)
    unless record.reply_status.present? && record.reply_status.length <= 1
      record.errors.add(:reply_status, REPLY_STATUS_ERROR_MSG)
    end
  end

  def validate_reply_status_inclusion(record)
    unless [ACCEPT_STATUS, REJECT_STATUS, WAITING_STATUS].include?(record.reply_status)
      record.errors.add(:reply_status, REPLY_STATUS_INCLUSION_ERROR_MSG)
    end
  end

  def validate_duplicate_invitation(record)
    conditions = {
      to_id: record.to_id,
      from_id: record.from_id,
      assignment_id: record.assignment_id,
      reply_status: record.reply_status
    }
    if Invitation.where(conditions).exists?
      record.errors[:base] << DUPLICATE_INVITATION_ERROR_MSG
    end
  end

  def validate_to_from_different(record)
    if record.from_id == record.to_id
      record.errors.add(:from_id, TO_FROM_SAME_ERROR_MSG)
    end
  end
end
```

**And then you would use this validator in your Invitation model like so:**

```
# app/models/invitation.rb
class Invitation < ApplicationRecord
  after_initialize :set_defaults

  belongs_to :to_user, class_name: 'User', foreign_key: 'to_id', inverse_of: false
  belongs_to :from_user, class_name: 'User', foreign_key: 'from_id', inverse_of: false
  belongs_to :assignment, class_name: 'Assignment', foreign_key: 'assignment_id'

  validates_with InvitationValidator
end
```

Reply:
Done! Now the invitation model looks simple and better.